### PR TITLE
WIP: burn-train in the browser

### PR DESCRIPTION
**Pull request by AlexErrant**
### Checklist

- [ ] Confirm that `run-checks` script has been executed.

This has never worked on my machine. Gonna lean on CICD for this.

### Related Issues/PRs

Closes https://github.com/burn-rs/burn/issues/921

### Changes

Here's part 1, where I replace a `spawn` in `EventStoreClient` with a webworker implementation pretty much copied straight from [here](https://www.tweag.io/blog/2022-11-24-wasm-threads-and-messages/). The total files changed is pretty large, but most of it is just a new example project I made called `examples/train-web`. That project uses PNPM and Vite, which stylistically diverges from the minimalistic `examples/mnist-inference-web`. I do so to more accurately reflect the usage in a real-world project that uses Typescript and NPM packages.

To make this PR easier to review, I'm following the atomic commit style and will refrain from force-pushing this branch.

### Testing

I have a branch [here](https://github.com/AlexErrant/burn/compare/train-browser..spawnWebWorkerTest) that demos `sender.send` working. Console output as follows:

```
train.js:325 INFO examples/train-web/train/src/lib.rs:17 
Hello from Rust
train.js:325 INFO burn-train/src/metric/store/client.rs:110 
Got msg: OnEventTrain(EndEpoch(1337))
train.js:325 INFO burn-train/src/metric/store/client.rs:110 
Got msg: OnEventValid(EndEpoch(314159))
train.js:325 INFO burn-train/src/metric/store/client.rs:110 
Got msg: End
```

### Next step

Either getting data loading of mnist working or exploring what it'll take to write to the filesystem.

### CODE REVIEW

**Review by nathanielsimard — COMMENTED** [review]
We need to add the instruction to install `cargo-watch`, otherwise there is an error.
I think an important step is have a way to import datasets easily, at least for the example to work.

**Comment by AlexErrant** [comment]
Here's Part 2. I load the MNIST database and create its data loaders. It was pretty straightforward and I made virtually no changes to Burn.

After this change, the browser console output is `panicked at 'failed to spawn thread: Error { kind: Unsupported, message: "operation not supported on this platform" }'`. I plan on working on that next. It's in `burn-core`'s `MultiThreadDataLoader`... so that will be fun. I plan on using the strategy demoed in Part 1.

Also, it turns out [FileSystemSyncAccessHandle](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle) - which I was planning on using for sync file system access - only provides sync access to a predetermined file. I'm noticing that Burn may want sync access to a directory. This increases the probability that I have to use Asyncify, though I'll try to avoid it.

**Comment by AlexErrant** [comment]
Here's part 3. Things get a little complicated. I ignored `MultiThreadDataLoader` for now (also, we can train without it or the filesystem - `learner.fit` runs as of this change.)

I realized that creating a web worker is inherently asynchronous. After all, you pass a url to the constructor! So the web worker won't spin up until after the next microtask - _after_ `learner.fit`'s completion.

Instead of arbitrarily spawning a web worker as needed, I create a thread/worker pool. Users should eagerly spin up [`navigator.hardwareConcurrency`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/hardwareConcurrency) number of web workers, then run training. I copied most of the pool code from [here](https://github.com/rustwasm/wasm-bindgen/blob/main/examples/raytrace-parallel/src/pool.rs). There are other forks [here](https://github.com/amethyst/web_worker/) and [here](https://github.com/squirrel-labs/web_worker) but I'm not using them because I don't need their features.

I optimized for the smallest delta between the copied source code and the "working" code. There's still some work to be done, but I'm pushing now for early feedback. Remaining work includes (but is not limited to) creating a task queue, having workers take tasks from the task queue, and cleaning up the pool upon completion.

[This branch](https://github.com/AlexErrant/burn/compare/interleaved..train-browser) demos EventStoreClient messages being interleaved with the main thread's training output, indicating that they run concurrently. Console output as follows:

```
examples/train-web/train/src/lib.rs:34 Hello from Rust
burn-train/src/learner/train_val.rs:125 Fitting Model[num_params=531178]
burn-train/src/learner/epoch.rs:100 Executing training step for epoch 1
burn-train/src/learner/epoch.rs:110 Iteration 1
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(EndEpoch(1))
burn-train/src/learner/epoch.rs:44 Executing validation step for epoch 1
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(EndEpoch(1))
burn-train/src/learner/epoch.rs:100 Executing training step for epoch 2
burn-train/src/learner/epoch.rs:110 Iteration 1
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(EndEpoch(2))
burn-train/src/learner/epoch.rs:44 Executing validation step for epoch 2
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(EndEpoch(2))
burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: End
examples/train-web/train/src/train.rs:138 Done!
```

Prior to this change, `Done!` would appear before `EventStoreClient got msg`.

<details><summary>Rejected workarounds</summary>

1. In-lining the worker script using a data urls, [which have opaque origins](https://html.spec.whatwg.org/multipage/workers.html#:~:text=data%3A%20URLs%20create%20a%20worker%20with%20an%20opaque%20origin). I'm using `SharedArrayBuffer` for the perf, but this requires [COEP/COOP](https://web.dev/articles/coop-coep), which in turn requires same-origin. I also don't know if this actually synchronously starts a web worker - didn't care to check. Kinda doubt it.

2. `URL.createObjectURL`, which (obviously) creates a URL.
</details>

(I am so, so far out of my depth 😅. Please LMK if you can think of a better architecture. Also I'm very uncertain about [this `Arc::new`](https://github.com/Tracel-AI/burn/pull/938/commits/9a355f28843eb7d0b99300bcd65291e81a7bb828#diff-6e526c2c551f703d3467a177f3f65cfb94af687fdf31a8e0b0f01ad2ea505118R165) and [this `Arc::new`](https://github.com/Tracel-AI/burn/pull/938/commits/9a355f28843eb7d0b99300bcd65291e81a7bb828#diff-6e526c2c551f703d3467a177f3f65cfb94af687fdf31a8e0b0f01ad2ea505118R173). Why am I forced to call it when the original code didn't call `Rc::New`? Does making `reclaim_slot` and `slot2` `mut` keep the same semanatics?)

**Comment by AlexErrant** [comment]
Nice! I was wondering if Burn "native" could use a worker pool, but didn't know if that would have any specific benefits. I'm unfamiliar with `ComputeServer` and trust your reasoning. [(Related discussion on WGPU server.)](https://github.com/Tracel-AI/burn/pull/833#discussion_r1396530058)

I am interested in keeping PRs small and subscribe to "merge early & often". I agree with eventually moving pool to another crate (I actually moved it to `burn-core` in an unpushed commit... then I discovered the async nature of web-worker constructors). Should we do that in subsequent PR, and focus this PR on making `spawn` production ready? That would involve adding a task queue/task de-queuing/pool clean-up. I see related crates about [threadpooling](https://docs.rs/threadpool/latest/threadpool/) and [task queues](https://docs.rs/work-queue/latest/work_queue/), though I've no idea if they're good/recommended/solve our problems/work with web-workers. Not sure if you have experience here (obviously I don't) but I'm game to keep working on this!

**Comment by nathanielsimard** [comment]
@AlexErrant sure we can work on making the `spawn` method work correctly before moving to refactoring. I'm not sure how we should go about making a task queue for wasm, so I guess you can "discover" it 😅

**Comment by AlexErrant** [comment]
So uh, [I deleted ~99% of my additions to Burn](https://github.com/tracel-ai/burn/pull/938/commits/6d926e806d7a02b103677f09988b5225fe908977) and just used [`wasm-bindgen-rayon`](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon) instead. It creates a web-worker pool that is used by rayon. (The majority of the other commits are me fighting CI trying to add nightly 😭)

Does it make sense to replace `std::thread::spawn` with [`rayon::spawn`](https://docs.rs/rayon/latest/rayon/fn.spawn.html)? If so, I don't think we need [the util file](https://github.com/tracel-ai/burn/pull/938/files#diff-14d33a01f43b446583cdef15bdaec761b07139286a5db07b0e68b3d5c9040d00) (and therefore don't need to lift it to burn-common).

[This branch](https://github.com/AlexErrant/burn/compare/train-browser..rayon-interleaved) demos EventStoreClient messages being interleaved with the main thread's training output, indicating that they run concurrently. Console output as follows:

```
train_web.js:333 INFO examples/train-web/train/src/lib.rs:28 Hello from Rust
train_web.js:333 INFO burn-train/src/learner/train_val.rs:125 Fitting Model[num_params=531178]
train_web.js:333 INFO burn-train/src/learner/epoch.rs:100 Executing training step for epoch 1
train_web.js:333 INFO burn-train/src/learner/epoch.rs:110 Iteration 1
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
train_web.js:333 INFO examples/train-web/train/src/train.rs:81 Training Progress: TrainingProgress { progress: Progress { items_processed: 2, items_total: 2 }, epoch: 1, epoch_total: 2, iteration: 1 }
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(EndEpoch(1))
train_web.js:333 INFO burn-train/src/learner/epoch.rs:44 Executing validation step for epoch 1
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
train_web.js:333 INFO examples/train-web/train/src/train.rs:85 Validation Progress: TrainingProgress { progress: Progress { items_processed: 2, items_total: 2 }, epoch: 1, epoch_total: 2, iteration: 1 }
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(EndEpoch(1))
train_web.js:333 INFO burn-train/src/learner/epoch.rs:100 Executing training step for epoch 2
train_web.js:333 INFO burn-train/src/learner/epoch.rs:110 Iteration 1
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
train_web.js:333 INFO examples/train-web/train/src/train.rs:81 Training Progress: TrainingProgress { progress: Progress { items_processed: 2, items_total: 2 }, epoch: 2, epoch_total: 2, iteration: 1 }
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventTrain(EndEpoch(2))
train_web.js:333 INFO burn-train/src/learner/epoch.rs:44 Executing validation step for epoch 2
train_web.js:333 INFO examples/train-web/train/src/train.rs:85 Validation Progress: TrainingProgress { progress: Progress { items_processed: 2, items_total: 2 }, epoch: 2, epoch_total: 2, iteration: 1 }
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(MetricsUpdate(MetricsUpdate { entries: [], entries_numeric: [] }))
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: OnEventValid(EndEpoch(2))
train_web.js:333 INFO burn-train/src/metric/store/client.rs:110 EventStoreClient got msg: End
train_web.js:333 INFO examples/train-web/train/src/train.rs:142 Done!
```

If everything looks sensible, we might be getting close to a reasonable point of merging. (I certainly want it to reduce any more conflicts with main/`runchecks.rs`) The remaining work includes replacing `spawn` in places like `MultiThreadDataLoader`, and continuing to iterate on the example project (persisting the trained model, using said model, reporting progress to UI via `MetricsRenderer`).

**Comment by nathanielsimard** [comment]
@AlexErrant I think it makes sense to use a library that allows to spawn new threads working on web assembly (web workers). Not sure if the crate that uses rayon is the best long-term solution, since it doesn't seem to be maintained much, but it can be a great solution in the short medium term, since we could alway change the `spawn` implementation with a new shinny tools that works with web assembly (the abstraction isn't leaky and invasive like `async` would be).

**Comment by AlexErrant** [comment]
A few words in praise of `wasm-bindgen-rayon`: It [has quite a few dependants](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/network/dependents), so should be pretty well battletested. The 3 open issues are pretty minor or are blocked by upstream, and the PR is explicitly left open for feedback. More importantly, its source code is very simple/straightforward: [lib.rs](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/main/src/lib.rs) is 147LOC and is pretty much a wrapper around `rayon::ThreadPoolBuilder::new()` and [workerHelpers.js](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon/blob/main/src/workerHelpers.js) is 108LOC and most of it is comments. It's certainly a lot better than what I was building in this PR.

What would you like to be done before merging? You said earlier

> If the code modified is clean enough in the existing burn crates, we could merge this even if the example isn't completed, and continue improving it incrementally with smaller PRs.

The burn code changes are clean. The example repo is a WIP, but is functional (well, if you include [some minor changes](https://github.com/AlexErrant/burn/compare/train-browser..rayon-interleaved)). There's still some work to be done outlined at the bottom of my previous comment; would you like some of them to be done in this branch? I wouldn't mind doing that - just trying to keep this PR small and as easily reviewable as possible :)

**Comment by nathanielsimard** [comment]
@AlexErrant Sorry for the slow reply. I think it would be better to send small PRs, since this one is still hard to review correctly. Also, Burn should run on Rust stable, so we may need to put feature flags inplace if training in the browser isn't possible on stable yet.

**Comment by AlexErrant** [comment]
I'm not sure how to split this PR into smaller pieces. [I made a video here](https://studio.youtube.com/video/QmOJfRnrJKg/edit) of me going through the changes and justifying them. Please watch it over lunch or something. Apologies for the crackly mic. If you have ideas on how to make this PR smaller, please LMK!

Here's a link about feature unification https://nickb.dev/blog/cargo-workspace-and-the-feature-unification-pitfall/

I skip over the example code because it's mostly a copy/paste of existing example code, just reworked for the web. I can go over that too if you want.

Around the 8-minute mark I say "obviously" clippy doesn't need to run on 4 projects. That was just me being nervous and forgetting - clippy just fails to run on those projects with the error:

```
[2023-12-23T03:09:32Z INFO  xtask::runchecks] /home/alex/burn/burn-wgpu $ cargo clippy --color=always --all-targets -- -D warnings

warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /home/alex/burn/examples/image-classification-web/Cargo.toml
workspace: /home/alex/burn/Cargo.toml
    Checking hashbrown v0.14.2
   Compiling serde_derive v1.0.192
   Compiling thiserror-impl v1.0.50
    Checking burn-common v0.12.0 (/home/alex/burn/burn-common)
    Checking rand_distr v0.4.3
   Compiling futures-util v0.3.29
   Compiling serial_test_derive v2.0.0
    Checking ndarray v0.15.6
    Checking indexmap v2.1.0
    Checking gpu-descriptor v0.2.4
    Checking burn-compute v0.12.0 (/home/alex/burn/burn-compute)
    Checking dashmap v5.5.3
    Checking thiserror v1.0.50
    Checking naga v0.14.1
    Checking futures-executor v0.3.29
    Checking futures v0.3.29
    Checking serial_test v2.0.0
    Checking serde v1.0.192
    Checking wgpu-hal v0.18.1
    Checking half v2.3.1
    Checking serde_json v1.0.108
    Checking burn-tensor v0.12.0 (/home/alex/burn/burn-tensor)
    Checking text_placeholder v0.5.0
    Checking wgpu-core v0.18.1
    Checking burn-autodiff v0.12.0 (/home/alex/burn/burn-autodiff)
    Checking burn-fusion v0.12.0 (/home/alex/burn/burn-fusion)
    Checking burn-ndarray v0.12.0 (/home/alex/burn/burn-ndarray)
    Checking wgpu v0.18.0
    Checking burn-wgpu v0.12.0 (/home/alex/burn/burn-wgpu)
error[E0599]: the function or associated item `new` exists for struct `ElemWiseBenchmark<Fusion<Wgpu>>`, but its trait bounds were not satisfied
  --> burn-wgpu/benches/fused_elemwise.rs:65:54
   |
11 | struct ElemWiseBenchmark<B: Backend> {
   | ------------------------------------ function or associated item `new` not found for this struct
...
65 |     run_benchmark(ElemWiseBenchmark::<Fusion<Wgpu>>::new(
   |                                                      ^^^ function or associated item cannot be called on `ElemWiseBenchmark<Fusion<Wgpu>>` due to unsatisfied trait bounds
   |
  ::: /home/alex/burn/burn-fusion/src/backend.rs:17:1
   |
17 | pub struct Fusion<B> {
   | --------------------
   | |
   | doesn't satisfy `_: ActivationOps<Fusion<Wgpu>>`
   | doesn't satisfy `_: Backend`
   | doesn't satisfy `_: BoolTensorOps<Fusion<Wgpu>>`
   | doesn't satisfy `_: IntTensorOps<Fusion<Wgpu>>`
   | doesn't satisfy `_: ModuleOps<Fusion<Wgpu>>`
   | doesn't satisfy `_: TensorOps<Fusion<Wgpu>>`
   |
note: trait bound `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::backend::Backend` was not satisfied
  --> burn-wgpu/benches/fused_elemwise.rs:11:29
   |
11 | struct ElemWiseBenchmark<B: Backend> {
   |        ---------------------^^^^^^^-
   |                             |
   |                             unsatisfied trait bound introduced here
   = note: the following trait bounds were not satisfied:
           `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::ops::ActivationOps<burn_fusion::Fusion<burn_wgpu::Wgpu>>`
           which is required by `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::backend::Backend`
           `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::ops::ModuleOps<burn_fusion::Fusion<burn_wgpu::Wgpu>>`
           which is required by `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::backend::Backend`
           `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::ops::IntTensorOps<burn_fusion::Fusion<burn_wgpu::Wgpu>>`
           which is required by `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::backend::Backend`
           `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::ops::BoolTensorOps<burn_fusion::Fusion<burn_wgpu::Wgpu>>`
           which is required by `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::backend::Backend`
           `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::ops::TensorOps<burn_fusion::Fusion<burn_wgpu::Wgpu>>`
           which is required by `burn_fusion::Fusion<burn_wgpu::Wgpu>: burn_tensor::backend::Backend`

error[E0277]: the trait bound `burn_wgpu::Wgpu: burn_fusion::FusionBackend` is not satisfied
  --> burn-wgpu/benches/fused_elemwise.rs:65:19
   |
65 |     run_benchmark(ElemWiseBenchmark::<Fusion<Wgpu>>::new(
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `burn_fusion::FusionBackend` is not implemented for `burn_wgpu::Wgpu`
   |
   = help: the trait `burn_tensor::backend::Backend` is implemented for `burn_fusion::Fusion<B>`
   = note: required for `burn_fusion::Fusion<burn_wgpu::Wgpu>` to implement `burn_tensor::backend::Backend`
note: required by a bound in `ElemWiseBenchmark`
  --> burn-wgpu/benches/fused_elemwise.rs:11:29
   |
11 | struct ElemWiseBenchmark<B: Backend> {
   |                             ^^^^^^^ required by this bound in `ElemWiseBenchmark`

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `burn-wgpu` (bench "fused_elemwise") due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

I'm not sure how to fix it, so I just ignored 'em. Notably they failed (on my machine) even without my changes, so I'm not sure it ever worked for those workspaces.

**Comment by nathanielsimard** [comment]
@AlexErrant Looking at the video:

1. We have to support Rust stable on the CI, we may also test on nightly or require nightly to run the example, but we can't switch it to nigtly by default.
2. For spawn: I would prefer to have a trait instead of `FnOnce`, it would be a be cleaner:

```rust
trait AsyncTask {
    fn join(self: Box<Self>) -> Result<(), AsyncTaskError>
}
pub type AsyncTaskBoxed = Box<dyn AsyncTask>;
```

Then you can use `AsyncTaskBoxed` where you need the type.

3. For the changes in `runchecks`, I think a lot of the things are "patched" to support nightly, but for simplification I would not change it for this PR. We have to think a bit harder to add a clean way to check parts of the framework that require nightly, and I prefer _not_ checking them at first instead of complicating the CI too much. It's also the kind of changes that is easier to review when it's the only thing that change in a PR.

**Comment by nathanielsimard** [comment]
@AlexErrant Thanks a lot for the video, it helps reviewing the PR. See my comments bollow:

1. We need to support Rust stable on the CI. While testing on nightly or requiring nightly to run the example is acceptable, switching to nightly by default is not an option.

2. Regarding `spawn`: I would prefer using a trait instead of `FnOnce` for a cleaner implementation:

```rust
trait AsyncTask {
    fn join(self: Box<Self>) -> Result<(), AsyncTaskError>;
}
pub type AsyncTaskBoxed = Box<dyn AsyncTask>;
```

You can then use `AsyncTaskBoxed` where the type is needed.

3. Concerning the changes in `runchecks`, many aspects seem "patched" to support nightly. However, for simplification, I suggest avoiding changes in this PR. We should carefully consider a clean way to check parts of the framework that require nightly. I prefer initially not checking them rather than complicating the CI too much. These changes are also easier to review when they are the only modifications in a PR.

While I appreciate your efforts on the CI, since I discovered several small issues, such as not adding `--workspace` to avoid feature unification, these can be addressed in a subsequent PR.

In conclusion, please avoid updating the CI for this PR. Just incorporate the example and the new way of launching threads.

**Comment by L-M-Sherlock** [comment]
How to build on this branch? I run this command:

```bash
RUSTFLAGS='-C target-feature=+atomics,+bulk-memory,+mutable-globals' cargo +nightly build --target wasm32-unknown-unknown -Z build-std=panic_abort,std
```

But I get these errors:
```
error: failed to run custom build command for `zstd-sys v2.0.9+zstd.1.5.5`
```
and
```
warning: zstd-sys@2.0.9+zstd.1.5.5: error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
```

**Comment by AlexErrant** [comment]
@L-M-Sherlock are you sure you're building `burn/examples/train-web/train`? Your command works for me there. (Though FWIW, I've been running [`./dev.sh`](https://github.com/tracel-ai/burn/pull/938/files#diff-a9d6b865c2c839f5e918652e57021ea2351500cb03c02dbfc97ad84421e936acR15).) When I run ` cargo tree | grep zstd` from that dir I get no results; but when I run it from the top-most `burn` dir I see its a dependency for `burn-tch`, which I've no intention of running in the browser.

@Luni-4 Trunk looks interesting. It has a [library mode](https://trunkrs.dev/#library-crate), which I think is more relevant for my purposes. Since I'm not building a Rust WASM application, I'm not entirely sure if Trunk provides significant advantages for my example project. It has two features: dev server and change detection. My example project currently uses Vite (since I'm (sigh) a boring webdev) which provides a dev server and change detection in JS. For Rust change detection, I'm using [cargo-watch](https://github.com/tracel-ai/burn/pull/938/files#diff-a9d6b865c2c839f5e918652e57021ea2351500cb03c02dbfc97ad84421e936acR9). An advantage of cargo-watch is that it can be used for non-wasm targets. It's also entirely optional. Vite serves as the build tool, and I don't think Trunk will simplify the CI process since the vast majority of the complexity is going to come from adding nightly.

**Comment by Luni-4** [comment]
@AlexErrant I think you were writing something similar to this [example](https://github.com/trunk-rs/trunk/tree/main/examples/vanilla) for what concerns the `wasm` part, as you are importing wasm dependencies in `burn/train` . Obviously this part cannot be useful for a JS-only application

**Comment by L-M-Sherlock** [comment]
> are you sure you're building `burn/examples/train-web/train`?

It works in building `burn/examples/train-web/train`. But the `./run-checks.sh` is in the top-most dir. I don't know how to pass the CI check.

**Comment by AlexErrant** [comment]
@Luni-4 I'm not sure I understand. Let me see if I can try to re-address your concerns.

> Is it possible to use [Trunk](https://trunkrs.dev/) to build this project? It might simplify the building phase and CI

I don't think Trunk will (significantly) simplify the building phase as that's handled by Vite. Trunk is meant for Rust web apps, and my web app is JS. CI requires nightly, which is where the majority of the complexity will come from. As far as I can tell the only thing Trunk will do for me is automatically download and install `wasm-bindgen`, which isn't that great a benefit as users will first need to download and install Trunk. So really we're just replacing one download with another.

> I think you were writing something similar to this [example](https://github.com/trunk-rs/trunk/tree/main/examples/vanilla?rgh-link-date=2024-02-03T19%3A59%3A36Z) for what concerns the wasm part, as you are importing wasm dependencies in burn/train

That example is a Rust web app, while my example project is JS. `burn/examples/train-web/train` is a library used by `burn/examples/train-web/web`, a JS web app. `burn/examples/train-web/train` depends on `burn/burn-train` which in turn depends on `wasm-bindgen-rayon`, but it's still just a library (not a Rust WASM _application_) at the end of the day. This difference  (library vs application) is the main reason why I'm unsure what benefits Trunk provides to my example project.

---

@L-M-Sherlock ah yeah, this PR currently fails CI (as evidenced by the many ❌s). I was working on getting CI passing, primarily by splitting up the build so feature unification doesn't occur, but ended up discarding all those changes when @nathanielsimard said "please avoid updating the CI for this PR". However, given "this PR requires changes to the CI because it adds the `wasm-bindgen-rayon` dependency which in turn requires nightly.", I believe I'll need to go back to fiddling with CI. Waiting for approval though to avoid wasting everyone's time :)

**Comment by Luni-4** [comment]
@AlexErrant 

My concerns were born from the fact that you are building `wasm` in your code, I can see a shell script which compiles your example for the `wasm32` target, so emitting WebAssembly as output.
Since `trunk` automatically manages `wasm-bindgen-cli` and it also offers JavaScript` interoperability, I thought it would have been possible to use that framework to simplify the building process, and thus solve some CI problems. That's why I had asked the question, to figure out if it was feasible, but since your use-case is different, that's fine for me.

**Comment by nathanielsimard** [comment]
This pull request has expanded to the extent that I am not comfortable merging it. It interacts with numerous parts of the framework, and I am concerned that merging it might introduce excessive technical debt and slow down the development of other areas of the framework.

I appreciate @AlexErrant for exploring the training of models on wasm, but I believe the scope is somewhat extensive for a single pull request. There is still valuable insight to be gained from this experiment, so I have created issues to track progress on better supporting wasm (#1256).

Ultimately, I believe we need to establish a robust web worker implementation that functions effectively with Rust stable before attempting to support complex workflows on wasm, aside from inference. This is the primary issue to address in enhancing our wasm support (#1250).

Thank you once again for your hard work. Not all pull requests are meant to be merged to add value to the project, and this one is a perfect example. Gaining knowledge and exploring new use cases is highly valuable, and I appreciate your contributions in that regard.

**Comment by mfranzs** [comment]
Hi - to confirm the current state of this, is `burn-train` not supported in the browser currently? That may be the cause of my issue in https://github.com/tracel-ai/burn/issues/1257. 

Thank you!

**Comment by nathanielsimard** [comment]
No, `burn-train` does not currently support wasm. Right now, if you want to train on wasm, you would need to write your own training loop. This is because `burn-train` extensively relies on the file system, multi-threading, and sqlite for dataset loading, with other functionalities that are not compatible with wasm.

**Comment by AlexErrant** [comment]
Just to make sure we're on the same page - this PR demoed `burn-train` working with just multi-threading (via web workers). Sqlite was not required for training to work. (I did use sql.js, but the data could have just as easily been provided via JSON/CSV/uint8array.) I didn't use the file system either, though I added a (trivial) [`to_bytes`](https://github.com/tracel-ai/burn/pull/938/files#diff-eb4a8b19dc5876ecc0b1b382bc79733bc48d4712d4ba70a736b8327226893929) method to the model since the only other way to do model saving is to disk.

As a closing thought, since we probably want atomics for communication between web workers, I don't think that'll be landing in stable for quite some time. Relevant issue: https://github.com/rust-lang/rust/issues/77839

**Comment by AlexErrant** [comment]
FYI I more or less finished a proof-of-concept of getting the non-trivial project https://github.com/open-spaced-repetition/fsrs-rs training in the browser https://github.com/AlexErrant/fsrs-browser

I git submoduled burn to get the above working https://github.com/AlexErrant/burn/tree/fsrs-browser I've no intention of maintaining it for anyone else's use; this message simply serves as an FYI so other people can copy me if they wish. In particular, fsrs-rs uses burn v0.11.1, and I don't intend to update that project unless fsrs-rs also updates.

Notably for me, training with ndarray is viable.
